### ISSUE TITLE
Requests intermittently result in 500, directory in /tmp not found

### ISSUE THREAD

**Issue by systemed** [phusion/passenger#2531]
Hi!

I'm seeing intermittent 500s from calls to my app (Ruby, Apache) before they even hit the app, with nothing appearing in the Apache logs. The failures seem to be correlated with looking for a Passenger instance that isn't there.

If I make a request every 10s from my home machine to the server, and report the result:

    Ok at 2024-03-14 08:43:27 +0000: 200
    Ok at 2024-03-14 08:43:37 +0000: 200
    Ok at 2024-03-14 08:43:47 +0000: 200
    Ok at 2024-03-14 08:43:57 +0000: 200
    Ok at 2024-03-14 08:44:08 +0000: 200
    Error 500 at 2024-03-14 08:44:23 +0000
    Ok at 2024-03-14 08:44:33 +0000: 200
    Ok at 2024-03-14 08:44:43 +0000: 200

And at the same time I run strace on the server with `ps auxw | grep -E 'sbin/(apache|httpd)' | awk '{print"-p " $2}' | xargs strace -F -t 2>&1 | grep agents`, I see:

```
[pid 1137295] 09:43:27 connect(46, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1127315] 09:43:35 connect(24, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137300] 09:43:37 connect(34, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137286] 09:43:47 connect(60, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1127899] 09:43:49 connect(22, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137297] 09:43:57 connect(40, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137296] 09:43:58 connect(22, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110 <unfinished ...>
[pid 1127907] 09:43:58 connect(46, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110 <unfinished ...>
[pid 1127890] 09:43:59 connect(51, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137292] 09:44:07 connect(57, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 974641] 09:44:18 connect(22, {sa_family=AF_UNIX, sun_path="/tmp/passenger.3v8gpcQ/agents.s/core"}, 110) = -1 ENOENT (No such file or directory)
[pid 974641] 09:44:18 connect(22, {sa_family=AF_UNIX, sun_path="/tmp/passenger.3v8gpcQ/agents.s/core"}, 110) = -1 ENOENT (No such file or directory)
[pid 974641] 09:44:18 connect(22, {sa_family=AF_UNIX, sun_path="/tmp/passenger.3v8gpcQ/agents.s/core"}, 110) = -1 ENOENT (No such file or directory)
     (same ENOENT line repeated another 76 times)
[pid 974641] 09:44:22 connect(22, {sa_family=AF_UNIX, sun_path="/tmp/passenger.3v8gpcQ/agents.s/core"}, 110) = -1 ENOENT (No such file or directory)
[pid 974641] 09:44:22 connect(22, {sa_family=AF_UNIX, sun_path="/tmp/passenger.3v8gpcQ/agents.s/core"}, 110) = -1 ENOENT (No such file or directory)
[pid 1137304] 09:44:33 connect(24, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137306] 09:44:43 connect(27, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137302] 09:44:45 connect(28, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137307] 09:44:49 connect(33, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1127306] 09:44:53 connect(25, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137287] 09:45:03 connect(23, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
[pid 1137308] 09:45:14 connect(23, {sa_family=AF_UNIX, sun_path="/tmp/passenger.kJhFbbR/agents.s/core"}, 110) = 0
```

i.e. the burst of ENOENTs corresponds exactly with the failed request. There's a long burst of them lasting just a few seconds, while the app is otherwise very lightly loaded with an access only every few seconds. (Ignore the 1-hour time difference, my home machine's in the UK and the server's in Germany :) ) 

Looking in `/tmp`, there's no sign of 3v8gpcQ, only of kJhFbbR:

    root@portet /etc/apache2 # ls -dla /tmp/passenger*
    drwxr-xr-x 5 root root 4096 Mar 14 09:45 /tmp/passenger.kJhFbbR

and passenger-config reports this as the only instance:

    root@portet /etc/apache2 # passenger-config list-instances --json
    [{"instance_dir": {"created_at": 1710391501, "created_at_monotonic_usec": 31816086526154, "major_version": 4, "minor_version": 0, "path": "/tmp/passenger.kJhFbbR"}, "instance_id": "sabmj1-efkD2q-z0hCZT", "integration_mode": "apache", "name": "j5g1mFnS", "passenger_version": "6.0.17", "server_software": "Apache/2.4.41 (Ubuntu) Phusion_Passenger/6.0.17", "watchdog_pid": 1127232, "core_pid": 1127235}]

Similarly passenger-status:

    Version : 6.0.17
    Date    : 2024-03-14 09:58:11 +0100
    Instance: j5g1mFnS (Apache/2.4.41 (Ubuntu) Phusion_Passenger/6.0.17)

The failures are erratic but seem to come in bursts - there'll be a few for a while and then nothing for hours.

I'm at a loss to see why it's occasionally resorting to this other non-existent instance. I may have done something stupid but any clues would be greatly appreciated.

----

System details:

- Ubuntu 20.04, no containerisation, x86_64 (running at Hetzner)
- Passenger 6.0.17 installed from Phusion APT repo
- Ruby 2.7, no version manager
- Apache 2.4.41 running with `PrivateTmp=false`

**Comment by CamJN** [comment]
The contents of `/tmp` are probably being cleaned-up by some process, try moving the passenger instance registry dir to another path. (https://www.phusionpassenger.com/docs/references/config_reference/apache/#passengerinstanceregistrydir)

**Comment by systemed** [comment]
My first attempt had been to keep it in `tmp` but with PrivateTmp=false, but this had the issue above.

I then moved it to `/var/run/passenger-instreg` and the same issue persisted.

I have now moved it to `/usr/local/share/passenger` and - so far - all appears to be good. 

I will refrain from posting my thoughts on systemd ;) but hopefully this will help anyone else encountering the same issue. Thanks for your help!
